components/relay/src/lib.rs

```diff
@@ -210,12 +239,177 @@ impl RelayClient {
+        // Check if host ends with ".{domain}"
+        // e.g., list has "google.com", host is "accounts.google.com"
+        if host.len() > domain.len()
+            && host.ends_with(domain.as_str())
+            && host.as_bytes()[host.len() - domain.len() - 1] == b'.'
+        {
+            return true;
+        }
+        // Check if domain is a subdomain of host
+        // e.g., list has "www.google.com", host is "google.com"
+        if domain.len() > host.len()
+            && domain.ends_with(host)
+            && domain.as_bytes()[domain.len() - host.len() - 1] == b'.'
+        {
+            return true;
+        }
+    }
+
+    false
+}
+
+/// Determines whether Firefox Relay should be offered for a given origin.
+///
+/// This function checks:
+/// 1. If the user has a valid Relay token (Relay is enabled)
+/// 2. If the origin is in the denylist (returns false if so)
+/// 3. If the origin is in the allowlist (returns true if so, false otherwise)
+///
+/// The allowlist and denylist are automatically fetched from Remote Settings.
+/// If the allowlist is empty or missing, Relay will not be shown (fail-safe default).
+///
+/// # Arguments
+/// * `relay_client` - The RelayClient instance (contains RS clients and auth token)
+/// * `origin` - The website origin (e.g., "https://example.com")
+///
+/// # Returns
+/// `true` if Relay should be shown for this origin, `false` otherwise
+///
+/// # Mobile App Usage
+/// Mobile apps should:
+/// 1. Create a `RelayClient` with the FxA access token and RemoteSettingsService
+/// 2. Call this function when user focuses on an email field
+/// 3. Show Relay UI if the function returns `true`
+///
+/// # Phase 1 Focus
+/// For the initial implementation, this function focuses on the scenario where
+/// the user has already enabled Relay (RelayClient has auth token).
+/// Future phases will handle unauthenticated scenarios.
+///
+/// # Examples
+/// ```kotlin
+/// // Kotlin example
+/// val remoteSettingsService = RemoteSettingsService(...)
+/// val relayClient = RelayClient(
+///     serverUrl = "https://relay.firefox.com",
+///     authToken = fxaToken,
+///     remoteSettingsService = remoteSettingsService
+/// )
+///
+/// if (shouldShowRelay(relayClient, "https://example.com")) {
+///     // Show Relay UI
+/// }
+/// ```
+#[uniffi::export]
+pub fn should_show_relay(relay_client: Arc<RelayClient>, origin: String) -> bool {
+    // Check if user has a relay token first
+    if !relay_client.has_relay_token() {
+        log::debug!("User has no relay token, not showing Relay");
+        return false;
+    }
+
+    // If no origin is provided, don't show Relay
+    if origin.is_empty() {
+        log::debug!("Empty origin, not showing Relay");
+        return false;
+    }
+
+    // Log the origin we're checking
+    log::debug!("Checking if Relay should be shown for origin: {}", origin);
+
+    // Fetch denylist from Remote Settings
+    let denylist = relay_client.rs_client.get_denylist();
```

Comment:
Can this function use the remote settings client directly instead?

For Phase 1 today, this could work (because we are building for existing Relay users), but this falls apart when we move to all users as part of Phase 2. Would it be more straight-forward to do it this way?

```suggestion
    let denylist = rs_client.get_denylist();
```